crates/uv-auth/src/middleware.rs

```diff
@@ -247,13 +240,17 @@ impl Middleware for AuthMiddleware {
             (retry_request, Some(response))
         };
 
-        // Check in the cache first
-        let credentials = self.cache().get_realm(
-            Realm::from(retry_request.url()),
-            credentials
-                .map(|credentials| credentials.to_username())
-                .unwrap_or(Username::none()),
-        );
+        // Check if there are credentials in the realm-level cache
+        let credentials = self
+            .cache()
+            .get_realm(
+                Realm::from(retry_request.url()),
+                credentials
+                    .as_ref()
+                    .map(|credentials| credentials.to_username())
+                    .unwrap_or(Username::none()),
+            )
+            .or(credentials);
```

Comment:
This one line change fixes the bug. If we had a username in the URL cache but no password in the realm cache we'd drop the password here.